### PR TITLE
fix(mesi): fetchConcurrent returns first success, not first result (issue #111)

### LINKED ISSUE #111: [mesi] fetchConcurrent returns the first result even if it is an error
## Problem

`fetchConcurrent` in `mesi/include.go:107-139` races `Src` and `Alt` in parallel and returns the **first** result that arrives on the channel — regardless of whether that result is a success or an error:

```go
resultChan := make(chan esiResponse, 2)
...
go runTask(token.Src)
go runTask(token.Alt)

result := <-resultChan  // first arrival wins, even if err != nil
close(doneChan)
return result.Data, result.IsEsiResponse, result.Error
```

Scenario that should succeed but fails today:

- `Src` returns HTTP 500 at ~50 ms.
- `Alt` returns HTTP 200 at ~80 ms.
- `fetchConcurrent` returns the error from `Src` and `Alt` is cancelled before its response can be consumed.

The test `TestFetchConcurrentPrimaryFailsAltSucceeds` only passes because of a carefully placed `time.Sleep(100 * time.Millisecond)` in the `/primary` handler — i.e., the test cheats so that alt arrives first. Remove the sleep and the test flips.

## Impact

- **Severity**: Medium (silent loss of availability that the `concurrent` fetch mode is specifically designed to provide)
- The point of `fetch-mode="concurrent"` is "give me whichever succeeds first". Today it is "give me whichever finishes first, error or not", which is strictly worse than `fetchFallback` in the failure case.

## Proposed solution

Wait for the first **successful** result. Fall back to the last error if both fail.

```go
func fetchConcurrent(token *esiIncludeToken, config EsiParserConfig) (string, bool, error) {
    if token.Alt == "" {
        return singleFetchUrlWithContext(token.Src, config, config.Context)
    }

    ctx, cancel := context.WithCancel(config.Context)
    defer cancel()

    type result struct {
        data          string
        isEsiResponse bool
        err           error
    }
    ch := make(chan result, 2)

    run := func(u string) {
        data, isEsi, err := singleFetchUrlWithContext(u, config, ctx)
        select {
        case ch <- result{data, isEsi, err}:
        case <-ctx.Done():
        }
    }

    go run(token.Src)
    go run(token.Alt)

    var lastErr error
    for i := 0; i < 2; i++ {
        r := <-ch
        if r.err == nil {
            cancel() // stop the other in-flight request
            // drain the other slot so the goroutine can exit
            go func() { <-ch }()
            return r.data, r.isEsiResponse, nil
        }
        lastErr = r.err
    }
    return "", false, lastErr
}
```

Notes:

- `defer cancel()` on both success and failure paths guarantees the second goroutine can exit.
- Draining the second slot avoids a goroutine blocking on `ch <- result{}` if the context is already done and the `select` branch races (the buffered channel of size 2 makes this bounded).

## Tests to write

Rewrite / add (do **not** rely on timing sleeps — use synchronization):

1. `TestFetchConcurrent_PrimaryErrorsImmediatelyAltSucceeds` — primary returns 500 synchronously; alt returns 200; assert result is the alt payload, no error, both URLs were hit exactly once.
2. `TestFetchConcurrent_FirstSuccessWins` — both succeed; alt is held on a channel until primary completes; assert primary's body is returned. Then swap (hold primary) and assert alt's body is returned.
3. `TestFetchConcurrent_BothFailReturnsError` — both return 500; assert non-nil error, content is empty.
4. `TestFetchConcurrent_CancelsLoser` — use per-request context inspection (handler closes a channel on entry, waits on `r.Context().Done()` before writing); assert that once one side succeeds, the other's `r.Context()` is cancelled within e.g. 200 ms.
5. Delete the `time.Sleep(100 * time.Millisecond)` hack in the existing `TestFetchConcurrentPrimaryFailsAltSucceeds` — synchronize via channels instead.

## Acceptance criteria

- [ ] `fetchConcurrent` returns the first successful result, not the first result
- [ ] Loser is cancelled via shared context
- [ ] No goroutine leaks (verified with `go test -race` and `goleak` if feasible)
- [ ] Tests are deterministic — no `time.Sleep` as synchronization
- [ ] Doc comment on `FetchMode == "concurrent"` updated to describe "first success wins"

### PR DESCRIPTION
**Pull request by s2x**
## Summary
- Fix `fetchConcurrent` to return the first **successful** result instead of the first result (which could be an error)
- Cancel the losing request via shared context to avoid goroutine leaks
- Remove `time.Sleep` hack from existing test (`TestFetchConcurrentPrimaryFailsAltSucceeds`)
- Add deterministic test (`TestFetchConcurrentPrimaryFailsImmediatelyAltSucceeds`) that doesn't rely on timing

## Problem
`fetchConcurrent` in `mesi/include.go:107-139` raced `Src` and `Alt` in parallel and returned the **first** result that arrived on the channel — regardless of whether that result was a success or an error.

Scenario that should succeed but failed:
- `Src` returns HTTP 500 at ~50 ms
- `Alt` returns HTTP 200 at ~80 ms
- `fetchConcurrent` returned the error from `Src` and `Alt` was cancelled before its response could be consumed

## Solution
- Loop through results until finding a successful one
- Cancel the losing request via `defer cancel()` and context propagation
- Handle context cancellation in the main loop to avoid goroutine leaks
- Use `ctx.Done()` instead of `doneChan` for cleaner cancellation

## Changes
- `mesi/include.go`: Rewrite `fetchConcurrent` to wait for first success
- `mesi/fetchUrl_test.go`: Add new test, remove `time.Sleep` hack

## Acceptance Criteria
- [x] `fetchConcurrent` returns the first successful result, not the first result
- [x] Loser is cancelled via shared context
- [x] No goroutine leaks (verified with `go test -race`)
- [x] Tests are deterministic — no `time.Sleep` as synchronization
- [x] Doc comment on `fetchConcurrent` updated to describe "first success wins"

Closes #111

### CODE REVIEW

File: mesi/fetchUrl_test.go [hunk 1]
```diff
@@ -503,6 +502,38 @@ func TestFetchConcurrentPrimaryFailsAltSucceeds(t *testing.T) {
 	}
 }
 
+func TestFetchConcurrentPrimaryFailsImmediatelyAltSucceeds(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/primary" {
```
Comment (s2x) [inline mesi/fetchUrl_test.go:507]: This comment describes the pre-fix behavior but will be misleading after merge. Remove or update to explain the test's purpose rather than describing a bug that no longer exists.

```go
// Test that fetchConcurrent waits for successful result even when primary fails immediately
```

File: mesi/include.go [hunk 2]
```diff
@@ -116,26 +118,37 @@ func fetchConcurrent(token *esiIncludeToken, config EsiParserConfig) (string, bo
...
 	}
+	defer cancel()
 
 	resultChan := make(chan esiResponse, 2)
-	doneChan := make(chan struct{})
 
 	runTask := func(url string) {
 		data, isEsiResponse, err := singleFetchUrlWithContext(url, config, ctx)
 		select {
 		case resultChan <- esiResponse{Data: data, IsEsiResponse: isEsiResponse, Error: err}:
```
Comment (s2x) [inline mesi/include.go:128]: If parent context (`config.Context`) is cancelled before any result arrives, `lastErr` is `nil`, returning `("", false, nil)`. This is unlikely in practice (HTTP requests would fail with context errors first), but could occur in a race. Consider returning a context error when `lastErr == nil`:

```go
case <-ctx.Done():
    cancel()
    if lastErr == nil {
        return "", false, ctx.Err()
    }
    return "", false, lastErr
```

**Review by s2x — COMMENTED** [review]
## Review Summary

Fix is correct - `fetchConcurrent` now returns first successful result instead of first result.

### Issues Found

**1. Outdated test comment** - See comment on fetchUrl_test.go

**2. Edge case: nil error on parent context cancellation** - See comment on include.go

### Verdict
Correct fix. Clean up test comment before merge. Edge case is low severity but worth considering.
